### PR TITLE
[AST] Increase size of type variable id field to 20 bits

### PR DESCRIPTION
**Pull request by xedin**
Currently `id` field is only 12 bits long which is a lot smaller
than `graph index` which is 28 bits and creates problems with
large collections of literals since `id` is going to overflow
pretty quickly (4096 elements max). Let's change it so both
fields have the same 20 bits which is big enough.

Resolves: [SR-10914](https://bugs.swift.org/browse/SR-10914)
Resolves: rdar://problem/51616307

### CODE REVIEW

**Comment by xedin** [comment]
@swift-ci please test

**Review by jckarter — APPROVED** [review]
Looks good to me.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

File: include/swift/AST/Types.h [hunk 1]
```diff
@@ -338,14 +338,14 @@ class alignas(1 << TypeAlignInBits) TypeBase {
 
   SWIFT_INLINE_BITFIELD_FULL(TypeVariableType, TypeBase, 64-NumTypeBaseBits,
```
Comment (jrose-apple) [inline include/swift/AST/Types.h]: You need to update this line too, to say how many bits are actually used. (That detects overflow.)

Comment (jrose-apple) [reply include/swift/AST/Types.h]: The previous version of the line was wrong, too. It should have just been `32+4+28`, with no subtraction. Now it should be `NumTypeBaseBits+20+4+20`.

Comment (xedin) [reply include/swift/AST/Types.h]: Ah alright, I'm going to adjust that, I though that this means that in total it uses 64 bits - what ever 20 bit for TypeBase and 44 for the rest, which kind of fits into `64 - NumTypeBaseBits`

Comment (xedin) [reply include/swift/AST/Types.h]: But it's definitely better to be explicit about it.

Comment (xedin) [reply include/swift/AST/Types.h]: Actually that should be just `20 + 4 + 20` since `C` in that template is extra bits :)

Comment (jrose-apple) [reply include/swift/AST/Types.h]: Oops, you're right. I didn't realize it was just the local bits.

Comment (xedin) [reply include/swift/AST/Types.h]: Alright, modified it to be explicit.

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please test source compatibility

**Comment by xedin** [comment]
@swift-ci please smoke test

**Comment by davezarzycki** [comment]
This change makes the TypeVariableType bitfield very sensitive to TypeBase bitfield experiments. Can we add a comment suggesting what one should do? Perhaps dial back the graph index to 19? If that's the advice, can we just automate it? For example:
```
@@ -341,18 +341,19 @@ protected:
     : NumPadBits,
     NumProtocols : 16
   );
 
-  SWIFT_INLINE_BITFIELD_FULL(TypeVariableType, TypeBase, 20+4+20,
+  SWIFT_INLINE_BITFIELD_FULL(TypeVariableType, TypeBase, 20+4,
     /// The unique number assigned to this type variable.
     ID : 20,
 
     /// Type variable options.
     Options : 4,
 
     ///  Index into the list of type variables, as used by the
     ///  constraint graph.
-    GraphIndex : 20
+    GraphIndex : NumPadBits;
+    static_assert(NumPadBits >= 16, "TypeVariableType bitfield needs refactoring")
   );
 
   SWIFT_INLINE_BITFIELD(SILFunctionType, TypeBase, NumSILExtInfoBits+3+1+2,
     ExtInfo : NumSILExtInfoBits,

```

**Comment by davezarzycki** [comment]
See #25818

**Pull request merged**